File: ontrack-service/src/main/java/net/nemerosa/ontrack/service/metrics/EntityCountMetrics.java
package net.nemerosa.ontrack.service.metrics;

import net.nemerosa.ontrack.model.metrics.OntrackMetrics;
import net.nemerosa.ontrack.repository.StatsRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.actuate.metrics.Metric;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.Collection;

@Component
public class EntityCountMetrics implements OntrackMetrics {

    private final StatsRepository repository;

    @Autowired
    public EntityCountMetrics(StatsRepository repository) {
        this.repository = repository;
    }

    @Override
    @Transactional(readOnly = true)
    public Collection<Metric<?>> metrics() {
        return Arrays.asList(
                new Metric<>("gauge.entity.project", repository.getProjectCount()),
                new Metric<>("gauge.entity.branch", repository.getBranchCount()),
                new Metric<>("gauge.entity.build", repository.getBuildCount()),
                new Metric<>("gauge.entity.promotionLevel", repository.getPromotionLevelCount()),
                new Metric<>("gauge.entity.promotionRun", repository.getPromotionRunCount()),
                new Metric<>("gauge.entity.validationStamp", repository.getValidationStampCount()),
                new Metric<>("gauge.entity.validationRun", repository.getValidationRunCount()),
                new Metric<>("gauge.entity.validationRunStatus", repository.getValidationRunStatusCount()),
                new Metric<>("gauge.entity.property", repository.getPropertyCount()),
                new Metric<>("gauge.entity.event", repository.getEventCount())
        );
    }

}

File: ontrack-service/src/main/java/net/nemerosa/ontrack/service/metrics/EntityCountMetrics.kt
package net.nemerosa.ontrack.service.metrics

import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.binder.MeterBinder
import net.nemerosa.ontrack.repository.StatsRepository
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
@Transactional(readOnly = true)
class EntityCountMetrics(
        private val repository: StatsRepository
) : MeterBinder {

    private fun MeterRegistry.count(
            name: String,
            countFn: StatsRepository.() -> Int
    ) {
        gauge(
                "ontrack.entity.$name",
                repository,
                { repository.countFn().toDouble() }
        )
    }

    override fun bindTo(registry: MeterRegistry) {
        registry.count("project", { projectCount })
        registry.count("branch", { branchCount })
        registry.count("build", { buildCount })
        registry.count("promotionLevel", { promotionLevelCount })
        registry.count("promotionRun", { promotionRunCount })
        registry.count("validationStamp", { validationStampCount })
        registry.count("validationRun", { validationRunCount })
        registry.count("validationRunStatus", { validationRunStatusCount })
        registry.count("property", { propertyCount })
        registry.count("event", { eventCount })
    }

}
